bridge/src/bridge/withdraw_relay.rs

```diff
@@ -219,74 +182,30 @@ impl<T: Transport> Stream for WithdrawRelay<T> {
 						)
 						.collect::<error::Result<Vec<_>>>()?;
 
-					let message_value_sufficient_payloads = messages
-						.iter()
-						.map(|message| {
-							message_value_sufficient_payload(
-								&app.home_bridge,
-								message
-							)
-						})
-						.map(|payload| {
-							app.timer.timeout(
-								api::call(&app.connections.home, home_contract.clone(), payload),
-								app.config.home.request_timeout)
-						})
-						.collect::<Vec<_>>();
-
-					info!("fetching whether message values are sufficent");
-					WithdrawRelayState::FetchMessageValueSufficient {
-						future: join_all(message_value_sufficient_payloads),
-						messages,
-						signatures,
-						block,
-					}
-				},
-				WithdrawRelayState::FetchMessageValueSufficient {
-					ref mut future,
-					ref messages,
-					ref signatures,
-					block
-				} => {
-					let message_value_sufficient = try_ready!(future.poll());
-					info!("fetching whether message values are sufficent complete");
-
-					let app = &self.app;
-					let home_contract = &self.home_contract;
-
 					let relays = messages.into_iter()
 						.zip(signatures.into_iter())
-						.zip(message_value_sufficient.into_iter())
-						// ignore those messages that don't have sufficient
-						// value to pay for the relay gas cost
-						.filter(|&((ref message, _), ref is_message_value_sufficient)| {
-							// TODO [snd] this is ugly.
-							// in the future ethabi should return a bool
-							// for `is_message_value_sufficient`
-							// since the contract function returns a bool
-							let is_sufficient = U256::from(is_message_value_sufficient.0.as_slice()) == U256::from(1);
-							if !is_sufficient {
-								info!("value in message is not sufficient to cover relay costs. ignoring. message: {}", message.0.to_hex());
-							}
-							is_sufficient
-						})
-						.map(|((message, signatures), _)| withdraw_relay_payload(&app.home_bridge, &signatures, message))
-						.map(|payload| TransactionRequest {
-							from: app.config.home.account,
-							to: Some(home_contract.clone()),
-							gas: Some(app.config.txs.withdraw_relay.gas.into()),
-							gas_price: Some(app.config.txs.withdraw_relay.gas_price.into()),
-							value: None,
-							data: Some(payload),
-							nonce: None,
-							condition: None,
-						})
-						.map(|request| {
+						.map(|(message, signatures)| {
+							let payload: Bytes = app.home_bridge.functions().withdraw().input(
+								signatures.iter().map(|x| Signature::v_from_bytes(x.0.as_slice())),
+								signatures.iter().map(|x| Signature::r_from_bytes(x.0.as_slice()).0),
+								signatures.iter().map(|x| Signature::s_from_bytes(x.0.as_slice()).0),
```

Comment:
`v_from_bytes`, `r_from_bytes` and `s_from_bytes` panic if bytes is not long enough

Reply:
also, you don't have to call `as_slice`. `&` is shorter and commonly used ;)

Reply:
i'll take more care here to handle the possible failures and either use `Result` or a panic with a proof (`q.e.d.`) it can't happen!

Reply:
`as_slice(x) == &x[..]`, `as_slice(x) != &x`, right?

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_slice

usually i prefer `&x[..]` as well.
i prefer `.as_slice(x)` when it's part of a `.` chain since it's only requires a suffix and not also a prefix.